### PR TITLE
[klx] add kunlunxin support

### PR DESCRIPTION
**Pull request by reiase**
## Motivation and Context

Add Kunlunxin support to DIOPI

## Description

Add basic ops for resnet50

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/kunlunxin/common/common.hpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+static xdnn::Context* get_raw_context(bool update_context) {
+    if (tls_raw_ctx_.get() == nullptr || update_context) {
+        tls_raw_ctx_.reset(xdnn::create_context(), xdnn::destroy_context);
+        assert(tls_raw_ctx_.get() != nullptr);
+        diopiStreamHandle_t pstream = nullptr;
+        diopiGetStream(context, &pstream);
+        assert(pstream != nullptr);
+        tls_raw_ctx_.get()->xpu_stream = (XPUStream)pstream;
+        if (tls_raw_ctx_.get()->dev().type() == xdnn::kXPU1) {
+            std::cout << "running in KunLun1" << std::endl;
```
Comment (caikun-pjlab) [inline impl/kunlunxin/common/common.hpp:74]: 调试语句建议加上DEBUG开关

File: impl/kunlunxin/common/common.hpp [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+xtorch_vec build_xtorch_vec(diopiSize_t size) {
+    // xdnn_pytorch::IntArrayRef tmp(size.data, size.len);
+    xtorch_vec res;
+
+    // IntArrayRef2IntVector(tmp, res);
+    int64_t length = size.len;
+    for (int i = 0; i < length; i++) {
+        if (i > 100) throw "bad size";
+        printf("size: %p, %d, data %ld, len: %ld\n", &size, i, size.data[i], size.len);
```
Comment (caikun-pjlab) [inline impl/kunlunxin/common/common.hpp:173]: 同上

File: impl/kunlunxin/common/common.hpp [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            printf("tensor is nullptr\n");
+        }
+        return {{0}, {0}, xdnn_pytorch::ScalarType::kfloat32, (void*)nullptr};
+    }
+
+    // NOTE: diopiTensor default construct func, storage_ is nullptr.
+    // if calling data(), Segmentation fault
+    // example: diopi_test/python/tests/test_last_error.py, diopiTensor()
+    diopiSize_t dsize;
+    diopiGetTensorShape(tensor, &dsize);
```
Comment (caikun-pjlab) [inline impl/kunlunxin/common/common.hpp:195]: 建议加上返回值判断，提前感知到错误

File: impl/kunlunxin/functions/basic_op.cpp [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    return diopiSuccess;
+}
+
+DIOPI_API diopiError_t diopiRandomInp(diopiContextHandle_t ctx, diopiTensorHandle_t inout, int64_t from, const int64_t* to, diopiGeneratorHandle_t gen) {
+    xdnn::Context* ctx_xpu = impl::kunlunxin::set_cur_ctx(ctx);
+    xdnn_pytorch::Tensor _inout = impl::kunlunxin::build_xtorch_tensor(inout);
+    int64_t _to = to != nullptr ? *to : 0x7fffffff;
+    xdnn_pytorch::Generator generator;
+
+    // TODO:
```
Comment (caikun-pjlab) [inline impl/kunlunxin/functions/basic_op.cpp:53]: TODO 加上对应的人和事项

**Review by caikun-pjlab — COMMENTED** [review]
1. 需要签署一下CLA
2. 目前代码合入有冲突，麻烦merge一下main

File: impl/kunlunxin/common/common.hpp [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    // NOTE: if dsize.len is 0, the tensor is actually a scalar.
+    // if (dsize.len == 0) {
+    //     NOT_SUPPORTED("tensor is empty");
+    //     return {{0}, {0}, xdnn_pytorch::ScalarType::kfloat32, (void*)nullptr};
+    // }
+
+    void* data = nullptr;
+    diopiGetTensorData(const_cast<diopiTensorHandle_t>(tensor), &data);
```
Comment (caikun-pjlab) [inline impl/kunlunxin/common/common.hpp:207]: 用const_cast 不是一个很好地实践，https://eel.is/c++draft/expr.const.cast#6

XdnnTorch里面能否支持 const void* 创建const tensor呢？

**Review by caikun-pjlab — APPROVED** [review]

File: impl/kunlunxin/common/common.hpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            printf("%ld ", i);
+        }
+        printf("\n");
+    }
+    return {xtorch_tensor_sizes, xtorch_tensor_strides, scalar_type, reinterpret_cast<void*>(data)};
+}
+
+template <typename T>
+decltype(auto) build_xtorch_tensorlist(T* tensors, int64_t numTensors) {
+    std::vector<xdnn_pytorch::Tensor> vecAtTensor;
```
Comment (yangbofun) [inline impl/kunlunxin/common/common.hpp:253]: vector内填值，在知道大小的情况下，最好先申请好空间，然后再遍历填充值。
std::vector<xdnn_pytorch::Tensor> vecAtTensor(numTensors);
for (size_t i = 0; i < numTensors; ++i) {
        vecAtTensor[i]=build_xtorch_tensor(tensors[i]);
    }

File: impl/kunlunxin/test/export_runtime.cpp
Comment (yangbofun) [inline impl/kunlunxin/test/export_runtime.cpp:1]: image
这里可以用这两个变量来管理，不建立大量的软连接。

File: impl/kunlunxin/common/common.hpp (hunk 6 above)
Comment (reiase) [reply impl/kunlunxin/common/common.hpp]: 已经修正

File: impl/kunlunxin/test/export_runtime.cpp
Comment (reiase) [reply impl/kunlunxin/test/export_runtime.cpp:1]: 已经修正，参考其他厂商的测试cmake，修正了软连接问题

File: impl/scripts/build_impl.sh [hunk 7]
```diff
@@ -70,6 +70,12 @@ case $1 in
     cmake .. -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DIMPL_OPT=supa -DCMAKE_BUILD_TYPE=Release -DTEST=ON
     make -j8
     ;;
+  kunlunxin)
+    mkdir -p build && cd build
+    XPURT_TOOLKIT_ROOT=${CURRENT_DIR}/../../xpu_toolchain/xpurt XDNN_TOOLKIT_ROOT=${CURRENT_DIR}/../../xpu_toolchain/xdnn XDNNTORCH_TOOLKIT_ROOT=${CURRENT_DIR}/../../xpu_toolchain/xdnn_pytorch cmake ..  \
+        -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DIMPL_OPT=kunlunxin -DTEST=${DIOPI_BUILD_TESTRT}
```
Comment (yangbofun) [inline impl/scripts/build_impl.sh:76]: 这里设置-DTEST=ON，我们弃用了DIOPI_BUILD_TESTRT。

Comment (reiase) [reply impl/scripts/build_impl.sh]: 已修复

**Review by yangbofun — APPROVED** [review]

**Pull request merged**